File: workspace/RainBackend/src/com/lucky/watisrain/backend/data/Route.java
package com.lucky.watisrain.backend.data;

import java.util.ArrayList;
import java.util.List;

/**
 * A Route is a journey from point A to point B. It is composed of a list of
 * RouteSteps, each of which represents a step of the journey.
 */
public class Route {

	private List<RouteStep> routeSteps;
	
	public Route(){
		routeSteps = new ArrayList<>();
	}
	
	public void addStep(RouteStep step){
		routeSteps.add(step);
	}
	
	public Location getStart(){
		return routeSteps.get(0).getStart();
	}
	
	public Location getEnd(){
		return routeSteps.get(routeSteps.size()-1).getEnd();
	}
	
	public List<RouteStep> getRouteSteps(){
		return routeSteps;
	}
	
	/*
	 * Sum of costs of all RouteSteps
	 */
	public double getTotalCost(){
		double sum = 0;
		for(RouteStep step : routeSteps){
			sum += step.getCost();
		}
		return sum;
	}
	
	
	/*
	 * A contracted route skips over passive locations, or if you go up multiple
	 * consecutive stairs.
	 */
	public Route getContractedRoute(){
		
		Route contractedRoute = new Route();
		
		int step = 0;
		Location lastNoteworthyNode = routeSteps.get(0).getStart();
		RouteStep routeToMerge = null;
		
		while(step < routeSteps.size()){
			
			// Precondition:
			// routeToMerge is empty, lastWorthyNode at active location
			
			while(step < routeSteps.size()){
				RouteStep cur_step = routeSteps.get(step);
				
				boolean is_different_building = !cur_step.getEnd().getPostion().equals(lastNoteworthyNode.getPostion());
				if(cur_step.getEnd().isPassive()) is_different_building = false;
				
				// If it's the first step, always merge, or we have a null route
				if(routeToMerge == null){
					routeToMerge = cur_step;
					step++;
				}
				
				// if start is passive but we're entering a building, add it and break
				else if(cur_step.getStart().isPassive() && is_different_building){
					routeToMerge = routeToMerge.mergeWith(cur_step);
					step++;
					break;
				}
				
				// if we're just entering a new building, don't add, just break
				else if(is_different_building){
					break;
				}
				
				// Add it and don't break
				else{
					routeToMerge = routeToMerge.mergeWith(cur_step);
					step++;
				}
			}
			
			contractedRoute.addStep(routeToMerge);
			lastNoteworthyNode = routeToMerge.getEnd();
			routeToMerge = null;

		}
		
		return contractedRoute;
	}
	
	
	public void printRouteToStdout(){
		double cumDistance = 0;
		for(RouteStep step : routeSteps){
			cumDistance += step.getCost();
			System.out.println(step + ". Total = " + (int)cumDistance);
		}
	}
	
}

File: iOS/WatBackend/WatBackend/Route.m

#import <Foundation/Foundation.h>
#import "Route.h"
#import "RouteStep.h"

@implementation Route

- (Route*) init{
    _routeSteps = [[NSMutableArray alloc] init];
    return self;
}

- (double) getTotalCost{
    double sum = 0;
    for(RouteStep *step in _routeSteps){
        sum += [step getCost];
    }
    return sum;
}

- (void) addStep:(RouteStep *)step{
    [_routeSteps addObject:step];
}

- (Location*)getStart{
    RouteStep *firstStep = [_routeSteps objectAtIndex:0];
    return [firstStep start];
}

- (Location*)getEnd{
    RouteStep *lastStep = [_routeSteps objectAtIndex:[_routeSteps count]-1];
    return [lastStep end];
}

- (Route*) getContractedRoute{
    NSMutableArray *marked = [[NSMutableArray alloc] init];
    for(int i=0; i<[_routeSteps count]-1; i++){
        
        RouteStep *rs_i = [_routeSteps objectAtIndex:i];
        RouteStep *rs_iplus1 = [_routeSteps objectAtIndex:i+1];
        Location *p1 = [rs_i start];
        Location *p2 = [rs_i end];
        Location *p3 = [rs_iplus1 end];
        
        if([p2 isPassive]){
            [marked addObject:[NSNumber numberWithInt:i]];
        }
        
        if([[p1 position] isEqual:[p3 position]]){
            [marked addObject:[NSNumber numberWithInt:i]];
        }
    }
    
    Route *contractedRoute = [[Route alloc] init];
    RouteStep *cur_step = nil;
    int step = 0;
    while(step < [_routeSteps count]){
        
        BOOL merge = [marked containsObject:[NSNumber numberWithInt:step]];
        
        if(cur_step == nil){
            cur_step = [_routeSteps objectAtIndex:step];
        }
        
        if(merge){
            cur_step = [cur_step mergeWith:[_routeSteps objectAtIndex:step+1]];
            step++;
        }
        else{
            [contractedRoute addStep:cur_step];
            cur_step = nil;
            step++;
        }
        
    }
    
    return contractedRoute;
}

@end
